freqtrade/freqtradebot.py

```diff
@@ -443,8 +443,22 @@ def execute_buy(self, pair: str, stake_amount: float, price: Optional[float] = N
 
         amount = stake_amount / buy_limit
 
-        order_id = self.exchange.buy(pair=pair, ordertype=self.strategy.order_types['buy'],
-                                     amount=amount, rate=buy_limit)['id']
+        order = self.exchange.buy(pair=pair, ordertype=self.strategy.order_types['buy'],
+                                  amount=amount, rate=buy_limit,
+                                  time_in_force=self.strategy.order_time_in_force['buy'])
+        order_id = order['id']
+        order_info = order.get('info', {})
+
+        # check if order is expired (in case of FOC or IOC orders)
+        # or rejected by the exchange.
+        order_status = order_info.get('status', '')
+        if order_status == 'EXPIRED' or order_status == 'REJECTED':
```

Comment:
i think this logic should be extracted to it's own function, which you'll also want to call regularily after getting the open order (to handle AON for example).

Also, are you sure the status-values are returned all uppercase? 
in handle-timeout `open` is received lowercase - which is aligned with what i get from ccxt when trying this manually.

Reply:
this would also make this easily testable ...

thinking out loud: are we sure we don't want to send a notification for "tried" buys?

Reply:
>i think this logic should be extracted to it's own function, which you'll also want to call regularily for open orders (to handle AON for example).

Agreed

>Also, are you sure the status-values are returned all uppercase?
in handle-timeout open is received lowercase - which is aligned with what i get from ccxt when trying this manually.

Well I tested it on Binance, and yes it is always uppercase and the feature is supported for binance only for now. 

>thinking out loud: are we sure we don't want to send a notification for "tried" buys?

I think this is a general problem that we have. currently buy signal leads to execute buy blindly (the only check is if we already have an open trade on the pair). there is no mechanism to control if the buy is not fulfilled even previously for GTC orders. if you put the interval at 15 min as an example every 5 second you receive the buy signal and the bot tries to buy blindly even if the market has moved against you ... 
I am going to open an issue on it as I think this is not the proper place to discuss about it. once we agree on the mechanism how to handle it then notifications come next ...

Reply:
@xmatthias let's discuss about it here https://github.com/freqtrade/freqtrade/issues/1335#issuecomment-444433418

Reply:
@xmatthias For extracting logic: I prefer doing it in a separated PR. as I am thinking about a Order class that handles the logic outside of freqtradebot.py. tell me if you are ok with that.

Reply:
i think extracting it would benefit as we can have dedicated, smallish unittests without a ton of mocking around ... which will facilitate and simplify refactoring later.

but it's nonblocking to me ... so your pick

Reply:
so I would pick a different PR finally :) as honestly I didn't find an easy way to do it.